lib/FIR/statements.h

```diff
@@ -54,6 +54,23 @@ class Stmt_impl {
   using StatementTrait = std::true_type;
 };
 
+// Some uses of a Statement should be constrained.  These contraints are imposed
+// at compile time.
+template<typename A = Stmt_impl> class QualifiedStmt {
+public:
+  QualifiedStmt() = delete;
+  template<typename B, std::enable_if_t<std::is_base_of_v<A, B>, int> = 0>
+  QualifiedStmt(Statement *stmt, const B &) : stmt{stmt} {}
```

Comment:
Should you have a dynamic check that the `*stmt` indeed points to variant containing B?

Or alternatively, you could create the statement on the go to ensure that stmt **is** a B. e.g:
`QualifiedStmt(BasicBlock*p, B &&b) : stmt{new Statement(p, std::forward<B>(b))} {}`

I see that you are only calling the constructor in `QualifiedInsert` and that this constructor would fit there. I would suggest to only allow such constructor if you can to avoid the risk of having `*stmt` and `B` mismatching.

Reply:
I'm not thrilled with this interface, but it gets past the compiler.  The plan is to just pass in the envelope and add enough template magic to type check.